docs/source/llms/rag/notebooks/question-generation-retrieval-evaluation.ipynb

```diff
@@ -504,95 +573,130 @@
     "\n",
-    "    question_list = []\n",
-    "    for q in response_queries.splitlines():\n",
-    "        q = \" \".join(q.split()[1:])\n",
-    "        question_list.append(q)\n",
+    "    messages = [{\"role\": \"user\", \"content\": prompt}]\n",
     "\n",
-    "    queries.append({\"chunk\": chunk, \"questions\": question_list})\n",
+    "    submit_function = {\n",
+    "        \"name\": \"submit_function\",\n",
+    "        \"description\": \"Call this function to submit the generated question and answer.\",\n",
+    "        \"parameters\": {\n",
+    "            \"type\": \"object\",\n",
+    "            \"properties\": {\n",
+    "                \"question\": {\n",
+    "                    \"type\": \"string\",\n",
+    "                    \"description\": \"The question asking for the key information in the given paragraph.\",\n",
+    "                },\n",
+    "                \"answer\": {\n",
+    "                    \"type\": \"string\",\n",
+    "                    \"description\": \"The answer to the question using the information in the given paragraph.\",\n",
+    "                },\n",
+    "            },\n",
+    "            \"required\": [\"question\", \"answer\"],\n",
+    "        },\n",
+    "    }\n",
     "\n",
-    "print(queries[0])"
+    "    response = openai.ChatCompletion.create(\n",
+    "        messages=messages,\n",
+    "        model=\"gpt-3.5-turbo\",\n",
+    "        functions=[submit_function],\n",
+    "        function_call=\"auto\",\n",
+    "        temperature=0.0,\n",
+    "        seed=SEED,\n",
+    "    )\n",
+    "    try:\n",
+    "        func_args = json.loads(response[\"choices\"][0][\"message\"][\"function_call\"][\"arguments\"])\n",
+    "        question = func_args[\"question\"]\n",
+    "        answer = func_args[\"answer\"]\n",
+    "        return question, answer\n",
+    "    except Exception as e:\n",
+    "        return str(e), \"N/A\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "844b40f2-533f-4ee5-9162-0091935cb8fd",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "queries = []"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {
+      "byteLimit": 2048000,
+      "rowLimit": 10000
+     },
+     "inputWidgets": {},
+     "nuid": "b88bd73d-222f-42c4-ada4-f251d40c2da2",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "n = len(filtered_df)\n",
+    "for i, row in filtered_df.iterrows():\n",
+    "    chunk = row[\"chunk\"]\n",
+    "    question, answer = generate_question_answer(chunk)\n",
+    "    print(f\"{i+1}/{n}: {question}\")\n",
+    "    queries.append(\n",
+    "        {\n",
```

Comment:
```
Alternatively, you can pin your installation to the old version, e.g. `pip install openai==0.28`
```

Got this error when running things.

I think it could be helpful to add pip installs at the top as well in case the user doesn't have an environment with langchain, bs4, seaborn, openai, mlflow, etc.

Reply:
Sure! Added a few.